### PR TITLE
add monero support

### PR DESCRIPTION
**Pull request by contrun**
Require the PR [Add monero signature support by contrun · Pull Request #1 · nervosnetwork/ed25519](https://github.com/nervosnetwork/ed25519/pull/1) from ed25519 repo. Currently ckb-auth-cli usages will be added in a separated PR.

### CODE REVIEW

File: docs/monero.md [hunk 1]
```diff
@@ -0,0 +1,171 @@
...
+```
+SigV23ZKheximq145tW14dshL17Jpqp2GJn296GfRnGqt3pMeaZU7xoEEAFr2Xm7Jc7xZjYWf6KhstZanA73to7uF6rea
+```
+Stripping prefix `SigV2`, this is the base64-encoded signature.
+
+## Verify the signature with `verify` subcommand
+```
+ckb-auth-cli monero verify -p a55ec8bb5b93aaffefd754996cb097228839aad6 -s 3ZKheximq145tW14dshL17Jpqp2GJn296GfRnGqt3pMeaZU7xoEEAFr2Xm7Jc7xZjYWf6KhstZanA73to7uF6rea -a 41eBLjYsK28CJD5z2b7FojMCDg6vERASShVZqAvnsC9LhS7saG8CmMo5Rm92wgnT8wa6nJVu57MHHjmnoyvTpCG7NQ7dErc -m spend
+-p a55ec8bb5b93aaffefd754996cb097228839aad6 -s 3ZKheximq145tW14dshL17Jpqp2GJn296GfRnGqt3pMeaZU7xoEEAFr2Xm7Jc7xZjYWf6KhstZanA73to7uF6rea
+```
```
Comment (XuJiandong) [inline docs/monero.md:49]: Why is the address "41eBLjYsK28CJD5z2b7FojMCDg6vERASShVZqAvnsC9LhS7saG8CmMo5Rm92wgnT8wa6nJVu57MHHjmnoyvTpCG7NQ7dErc" is required?

duplicated?
```
-p a55ec8bb5b93aaffefd754996cb097228839aad6 -s 3ZKheximq145tW14dshL17Jpqp2GJn296GfRnGqt3pMeaZU7xoEEAFr2Xm7Jc7xZjYWf6KhstZanA73to7uF6rea
```

File: c/auth.c [hunk 2]
```diff
@@ -363,6 +369,137 @@ int validate_signature_cardano(void *prefilled_data, const uint8_t *sig,
...
+
+    CHECK2(msg_len == BLAKE2B_BLOCK_SIZE, ERROR_INVALID_ARG);
+    CHECK2(sig_len == MONERO_DATA_SIZE, ERROR_INVALID_ARG);
+
+    uint8_t *mode_ptr = (uint8_t *)sig + MONERO_SIGNATURE_SIZE;
+    CHECK2(*mode_ptr != 0 || *mode_ptr != 1, ERROR_INVALID_ARG);
+
+    uint8_t *spend_pubkey = mode_ptr + sizeof(*mode_ptr);
+    uint8_t *view_pubkey = spend_pubkey + MONERO_PUBKEY_SIZE;
+    uint8_t *pubkey = *mode_ptr == 0 ? spend_pubkey : view_pubkey;
```
Comment (XuJiandong) [inline c/auth.c:481]: Why does view key can validate the signature?

File: docs/monero.md [hunk 3]
```diff
@@ -0,0 +1,171 @@
...
+is to sign the transaction hash (or maybe `sighash_all`, i.e. hashing all fields 
+including transaction hash and other witnesses in this input group)
+with `monero-wallet-cli`, and then leverage ckb-auth to check the validity of this signature.
+See [the docs](./auth.md) for more details.
+
+# Generate and verify transaction with ckb-auth-cli
+
+## Get the pub key hash with `parse` sub command.
+Here the argument given to `-a` is the address of monero account, while the argument given to `-m` refers to
+whether the spend key or view key is used to sign the transaction.
```
Comment (XuJiandong) [inline docs/monero.md:16]: Why can view key sign a transaction?

**Review by XuJiandong — CHANGES_REQUESTED** [review]

File: .gitmodules [hunk 4]
```diff
@@ -14,4 +14,4 @@
 
 [submodule "deps/ed25519"]
 	path = deps/ed25519
-	url = https://github.com/nervosnetwork/ed25519.git
```
Comment (XuJiandong) [inline .gitmodules:17]: If we need a update, push it to `nervosnetwork/ed25519.git` as a new branch. Then refer it here.

**Review by XuJiandong — CHANGES_REQUESTED** [review]

File: c/auth.c (hunk 2 above)
Comment (contrun) [reply c/auth.c:481]: The official `monero-wallet-cli` supports both signing with spend key and view key.

`monero-wallet-cli --wallet-file ckb-auth-test-wallet --password pw --offline help sign` outputs

```
 help sign
Command usage:
  sign [<account_index>,<address_index>] [--spend|--view] <filename>

Command description:
  Sign the contents of a file with the given subaddress (or the main address if not specified)
```
where `--spend/--view` is used to choose which key to sign the transaction.

I updated the documentation to reflect that.

File: docs/monero.md (hunk 3 above)
Comment (contrun) [reply docs/monero.md:16]: Ditto.

File: docs/monero.md (hunk 1 above)
Comment (contrun) [reply docs/monero.md]: Yes. This is duplicated. I removed that.

File: .gitmodules (hunk 4 above)
Comment (contrun) [reply .gitmodules:17]: When https://github.com/nervosnetwork/ed25519/pull/1 is merged, I will update this PR to use `ed25519` repo from `nervosnetwork`.

File: c/auth.c (hunk 2 above)
Comment (XuJiandong) [reply c/auth.c:481]: I don't think view key can transfer assets although it can sign transactions.

Comment (contrun) [reply c/auth.c:481]: It is also my understanding.

**Review by XuJiandong — APPROVED** [review]

**Pull request merged**